feat(client): add usage reporting

62w71st opened the pull request.

62w71st committed:
Pass in serverName to VpnTunnelService

62w71st committed:
Pass in serverName to VpnTunnelService

62w71st committed:
Merge branch 'master' of https://github.com/62w71st/outline-apps

62w71st committed:
merge

62w71st committed:
merge

62w71st committed:
add servername to cordova and electron

Commit:
change the origin of the outline plugin code

Commit:
conditionally pass serverName for Cordova only on Android.

Commit:
lint

62w71st committed:
merge with upstream

62w71st committed:
resolve conflicts

62w71st committed:
conflict resolution part 2.

Commit:
fix compilation errors

Commit:
Merge branch 'master' of https://github.com/62w71st/outline-apps

Commit:
Add serverName to logs and remove getServerName

Commit:
remove ongoing

62w71st committed:
Merge branch 'master' into master

62w71st committed:
Merge branch 'master' of https://github.com/Jigsaw-Code/outline-apps

62w71st committed:
Add reporting client.

62w71st committed:
usage-reporting parser.

62w71st committed:
Merge remote-tracking branch 'upstream/master' into usage-reporting-v2

62w71st committed:
Merge branch 'Jigsaw-Code:master' into usage-reporting-v2

62w71st committed:
Merge branch 'master' of https://github.com/Jigsaw-Code/outline-apps into usage-reporting-v2

62w71st committed:
Merge branch 'usage-reporting-v2' of https://github.com/62w71st/outline-apps into usage-reporting-v2

62w71st committed:
GET -> POST

62w71st committed:
Merge branch 'master' of https://github.com/Jigsaw-Code/outline-apps into usage-reporting-v2

62w71st committed:
Add OnConnected() callback.

62w71st committed:
fix build

62w71st committed:
Add connect and disconnect to the outline client

62w71st committed:
fix build errors

62w71st committed:
Merge branch 'master' of https://github.com/Jigsaw-Code/outline-apps into usage-reporting-v2

62w71st committed:
pass client as streamDialer to reporting client

62w71st committed:
Merge remote-tracking branch 'upstream/master' into usage-reporting-v2

62w71st committed:
add parsing 1/2

62w71st committed:
create usagereporter in the client

62w71st committed:
merge changes from upstream master

62w71st committed:
add session client

62w71st committed:
Fix

62w71st committed:
fix build errors in OutlineAndroidLib

62w71st committed:
Add usage report config to ts and java files

62w71st committed:
Separate the session client from the transport client.

62w71st committed:
fix build error.

62w71st committed:
export getters in sessionclient for gobind to export SessionClient!

62w71st committed:
Fix build.

62w71st committed:
Fix Getter.

62w71st committed:
gobind

fortuna commented on client/go/outline/config/config_usage_reporter.go:
Move to client/go/usagereport to keep the application packages simpler

jyyi1 commented on client/go/outline/config/config_tcpudp.go:
```diff
@@ -24,8 +24,8 @@ import (
 type TCPUDPConfig struct {
-	TCP ConfigNode
-	UDP ConfigNode
+	TCP           ConfigNode
```
Revert the change? `gofmt` should automatically remove the spaces?

jyyi1 commented on client/go/outline/reporting/cookies.json:
```diff
@@ -0,0 +1 @@
+{"https://example.com:443":[{"name":"client-id","value":"random_client_id","domain":"","path":""}]}
```
I don't think we need to check in this file?

jyyi1 commented on client/src/cordova/android/OutlineAndroidLib/outline/src/main/aidl/org/outline/TunnelConfig.aidl:
```diff
@@ -18,4 +18,5 @@ parcelable TunnelConfig {
   String id;
   String name;
   String transportConfig;
+  String usageReportConfig;
```
This can be combined with `transportConfig`, and get parsed in Go? So we don't need to touch the Java code.

jyyi1 commented on client/src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -237,6 +242,7 @@ private synchronized PlatformError startTunnel(
         isAutoStart ? tunnelStore.isUdpSupported() : udpConnError == null;
     try {
       final PlatformError tunError = vpnTunnel.connectTunnel(client, remoteUdpForwardingEnabled);
+      sessionClient.start();
```
Call the usage report logic in the Go code called by the `connectTunnel` above.

jyyi1 commented on client/src/cordova/plugin/android/java/org/outline/OutlinePlugin.java:
```diff
@@ -209,7 +209,8 @@ private void executeAsync(
           final String serverName = args.getString(1);
           final String transportConfig = args.getString(2);
-          sendActionResult(callback, startVpnTunnel(tunnelId, transportConfig, serverName));
+          final String usageReportConfig = args.getString(3);
```
No need to have a separate config, it's already included in the `transportConfig`.

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,271 @@
+	}
+
+	// Save the cookies data to a file
+	file, err := os.Create(filename)
```
Where will this file get saved to in Android? Application data?

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,271 @@
+	ticker := time.NewTicker(ur.Interval)
+	defer ticker.Stop()
+
+	for {
```
Is this dead loop? Shall we have some logic for shutting down the reporting after user disconnects?

jyyi1 replied:
If we have to, we might need to avoid create multiple dead loops when user connects to different servers?

62w71st replied:
Done.

jyyi1 reviewed (commented):
Thanks for the contribution! As discussed in the meeting, please:

- Refactor the [`newTunnel`](https://github.com/Jigsaw-Code/outline-apps/blob/7d252123109f5ea33207e6beac9732be3d9dd133/client/go/outline/tun2socks/tunnel.go#L67-L85) to accept a `Client` parameter
- Parse the reporting config in [`NewClientWithBaseDialers`](https://github.com/Jigsaw-Code/outline-apps/blob/7d252123109f5ea33207e6beac9732be3d9dd133/client/go/outline/client.go#L63), we can add new reporting field to `type Client struct`
- Start the reporting [here](https://github.com/Jigsaw-Code/outline-apps/blob/7d252123109f5ea33207e6beac9732be3d9dd133/client/go/outline/tun2socks/tunnel.go#L84) or [here](https://github.com/Jigsaw-Code/outline-apps/blob/7d252123109f5ea33207e6beac9732be3d9dd133/client/go/tunnel/tunnel.go#L61) by reading the parsed reporting config field in `Client`. Maybe the latter one is better because we will add "Stop reporting" logic in the same file, but this means we need to pass additional reporting config parameter to `Tunnel`, this is a trade-off.
- Stop the reporting [here](https://github.com/Jigsaw-Code/outline-apps/blob/7d252123109f5ea33207e6beac9732be3d9dd133/client/go/tunnel/tunnel.go#L44)

Please also fix the go tests, the CI seems to be failing.

62w71st committed:
Merge branch 'Jigsaw-Code:master' into usage-reporting-v2

62w71st committed:
Addressed a few comments and removed usage reporting configuration from the
// Android plugin.

62w71st committed:
Removed usage reporting config from TunnelConfig.aidl

62w71st committed:
Merge branch 'master' of https://github.com/Jigsaw-Code/outline-apps into usage-reporting-v2

62w71st committed:
remove cookies.json

62w71st committed:
merge upstream

62w71st committed:
Fix tests.

62w71st committed:
Pass in session_report to the cordova plugin.

62w71st committed:
Fix module_test.go tests.

62w71st committed:
stop reporting when tunnel is stopped.

62w71st reviewed (commented):
PTAL

62w71st committed:
Merge branch 'master' into usage-reporting-v2

jyyi1 commented on client/src/cordova/android/OutlineAndroidLib/outline/src/main/aidl/org/outline/IVpnTunnelService.aidl:
```diff
@@ -31,6 +31,7 @@ interface IVpnTunnelService {
    * extra with the result of the operation, as defined in OutlinePlugin.ErrorCode. Displays a
    * persistent notification for the duration of the tunnel.
    *
+   * @param serverName the server to which this tunnel connects.
```
revert?

```suggestion
```

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,274 @@
+)
+
+func CheckTCPConnectivity(tcp transport.StreamDialer, reportingUrl string) error {
+	tcpErr := connectivity.CheckTCPConnectivityWithHTTP(tcp, reportingUrl)
+	if tcpErr != nil {
+		return tcpErr
+	}
+	return nil
```
```suggestion
	return connectivity.CheckTCPConnectivityWithHTTP(tcp, reportingUrl)
```

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,274 @@
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+func CheckTCPConnectivity(tcp transport.StreamDialer, reportingUrl string) error {
```
Do we need to export this function?
```suggestion
func checkTCPConnectivity(tcp transport.StreamDialer, reportingUrl string) error {
```

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,274 @@
+	}
+
+	headers := string(response[:headersEnd])
+	fmt.Println("HTTP Headers:")
+	fmt.Println(headers)
```
Similar to above, debugging or logging? Please remove if it's for debugging, or use `slog` package if it's for logging.

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,274 @@
+						Name:  cookieParts[0],
+						Value: cookieParts[1],
+					})
+					fmt.Println("Cookie found:", cookie)
```
logging/debugging?

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,274 @@
+func loadCookies(jar http.CookieJar, filename string) ([]*url.URL, error) {
+	// Read cookies from the file
+	file, err := os.Open(filename)
+	if os.IsNotExist(err) {
```
According to [Go doc](https://pkg.go.dev/os#IsNotExist):

> This function predates [errors.Is](https://pkg.go.dev/errors#Is). It only supports errors returned by the os package. New code should use errors.Is(err, fs.ErrNotExist).

```suggestion
	if errors.Is(err, fs.ErrNotExist) {
```

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,274 @@
+	}
+
+	// Save the cookies data to a file
+	file, err := os.Create(filename)
```
As we discussed for privacy, do we need to allow the user to delete this cookie file (though removing the key or any other methods)?

emohandesi replied:
The cookie can be deleted when the key is deleted or when the Outline app is uninstalled. Do you want to add the option to be able to delete cookies manually?
@62w71st @fortuna

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,274 @@
+			}
+		case <-ctx.Done():
+			// Stop reporting when the context is canceled
+			fmt.Println("Stopping reporting...")
```
debugging?

jyyi1 commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,274 @@
+		case <-ticker.C:
+			err := Report(tcp, ur.Url)
+			if err != nil {
+				fmt.Printf("Report failed: %v\n", err)
```
Plz use `slog` for logging?

jyyi1 reviewed (commented):
Thanks for the update! The code looks a lot better.

fortuna commented on client/src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -317,10 +320,17 @@ private synchronized PlatformError startTunnel(
+
+    // Stop reporting on the Client instance
+    if (this.outlineClient != null) {
+        this.outlineClient.stopReporting();
```
Move to tearDownActiveTunnel. But I wonder if we should just implement that in stopRemoteDevice instead and revert this file. Why would we not align them?

fortuna commented on client/go/outline/vpn_linux.go:
```diff
@@ -27,6 +27,7 @@ import (
 type vpnConfigJSON struct {
 	VPNConfig       vpn.Config `json:"vpn"`
 	TransportConfig string     `json:"transport"`
+	SessionConfig   string     `json:"session_report"`
```
The whole `vpnConfigJSON` is the session config
```suggestion
	ReportConfig   string     `json:"report"`
```

fortuna commented on client/go/outline/client.go:
```diff
@@ -41,6 +45,10 @@ func (c *Client) ListenPacket(ctx context.Context) (net.PacketConn, error) {
 	return c.pl.ListenPacket(ctx)
 }
 
+func (c *Client) GetUsageReporter() *config.UsageReporter {
```
Why is this needed? You already provide Start/StopReporting. Remove?

fortuna commented on client/go/outline/client.go:
```diff
@@ -49,18 +57,35 @@ type NewClientResult struct {
+
 // NewClient creates a new Outline client from a configuration string.
-func NewClient(transportConfig string) *NewClientResult {
+func NewClient(transportConfig string, sessionConfig string) *NewClientResult {
```
This is not scalable. Please pass a single config. Rename transportConfig to clientConfig.

Same in the calls below.

fortuna replied:
Atually, use transportConfig, not clientConfig, since we already have that.

fortuna commented on client/go/outline/client.go:
```diff
@@ -49,18 +57,35 @@ type NewClientResult struct {
+
 // NewClient creates a new Outline client from a configuration string.
-func NewClient(transportConfig string) *NewClientResult {
+func NewClient(transportConfig string, sessionConfig string) *NewClientResult {
```
This is not a session config. It's a report config. But we should merge them into a client config.

fortuna commented on client/go/outline/client.go:
```diff
@@ -49,18 +57,35 @@ type NewClientResult struct {
 }
 
-func NewClientWithBaseDialers(transportConfig string, tcpDialer transport.StreamDialer, udpDialer transport.PacketDialer) (*Client, error) {
+func NewClientWithBaseDialers(transportConfig string, sessionConfig string, tcpDialer transport.StreamDialer, udpDialer transport.PacketDialer) (*Client, error) {
```
Same

fortuna commented on client/go/outline/client.go:
```diff
@@ -101,5 +126,31 @@ func NewClientWithBaseDialers(transportConfig string, tcpDialer transport.Stream
 	}
 
-	return &Client{sd: transportPair.StreamDialer, pl: transportPair.PacketListener}, nil
+	usageReportYAML, err := config.ParseConfigYAML(sessionConfig)
```
I think the easiest is to move the parsing of the clientConfig to the config module, which already takes care of this. And it allows you to use first_supported

62w71st committed:
Merge branch 'Jigsaw-Code:master' into usage-reporting-v2

62w71st commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,274 @@
+}
+
+// StartReporting calls the Report function at every internal.
+func StartReporting(ctx context.Context, tcp transport.StreamDialer, ur *config.UsageReporter) {
```
Current, cookies for all keys belonging to a provider are sent in the request. The provider should be able to dedupe them. Can you explain the requirement that motivates separating these cookies per key?

fortuna replied:
Each access key must have their own separate cookie jar to prevent possible cross-service leaks. It's a security requirement. Imagine two separate services talking to the same reporting service, for example.

Also, you need to delete the cookies when you delete the key, and this makes it easier, because you don't know the domains it needs to talk to.

emohandesi replied:
In order to prevent possible cross-service leaks, I am storing the cookie value received from the server along with the key ID (server ID) as a cookie key/value pair (`LocalKeyID=CookieValueReceivedFromReportingServer`).
Please note that the cookie received from the server is `client_id=CookieValueReceivedFromReportingServer`; however, since `client_id` is always the same and not needed, it is replaced by the `LocalKeyID` (server ID). When the client wants to retrieve the cookie of a key (server), it searches for the key ID in the cookies and sends the value of that key ID to the reporting server. Therefore, cross-service leaks are prevented.

Cookies are also deleted when the key (server) is removed from the client. Please note that in this design, which was explained in the previous paragraph, even if the cookie is not removed when the key is removed, no other reporting server will be able to access that cookie because they will never have the same local key ID.

62w71st committed:
Apply comments.

62w71st committed:
Merge branch 'usage-reporting-v2' of https://github.com/62w71st/outline-apps into usage-reporting-v2

62w71st committed:
Apply comments.

fortuna commented on client/src/www/app/outline_server_repository/vpn.cordova.ts:
```diff
@@ -28,7 +28,8 @@ export class CordovaVpnApi implements VpnApi {
       request.id,
       request.name,
-      request.config.transport
+      request.config.transport,
```
Resolved in my PR. Please pull the changes.

fortuna commented:
Please pull the latest master

62w71st committed:
Merge remote-tracking branch 'upstream/master' into usage-reporting-v2

62w71st committed:
This fix invloves the following changes:

1. Use http.client when contacting the reporing server.
2. Removes the cookie when its key ID (server ID) is removed from the app.
3. Tried to fix the logic of reading the `session_report` section of the key. However, this change is not complete. You can drop the changes made in client/go/outline/parse.go if you do not want to continue that and start from scratch.
4. refactoring NewClient to resolve the build errors
5. replace reporting URL as the cookie identifier with key ID (server ID)
6. transportConfig -> clientConfig
7. add reporting client test
8. use report instead of reporting_config to start the reporting block of the config

Tests:
Tests were done when the session_report section of the key was being read.
Cookie tests done after creating a new key that has a reporting server are as follows. All tests have been verified by adding print statements to the code.
		a. Start a connection to server 1 which has reporting URL 1, for the 1st time.
		i. I see a cookie received from the server and stored in the cookie jar and saved to cookies.json.
	b. Start a connection to the server for the 2nd time.
		i. I see the saved cookie is sent from Outline to the reporting server and no cookie is received from the reporting server.
	c. Start a connection server 2 that has a reporting server.
		i. I see a cookie received from the server and stored in the cookie jar and saved to cookies.json in addition to the previous cookie.
	d. Start a connection to server 2 for the 2nd time.
		i. I see the saved cookie is sent from Outline to the reporting server and no cookie is received from the reporting server.
	e. Remove a key from the Outline that has a reporting server.
		i. I see the corresponding cookie of the key is removed from cookies.json.
	f. Remove a key that does not have a reporting server.
		i. The code for cookie removal is run, but since the key ID is not found, no cookie is removed and I see the other cookies remain intact because when tried to connect the other two servers, I see they send their cookie and no cookie is sent to them in response.
	g. Start a connection to server 3 that does not have a reporting URL
		i. Reporting procedures are exited because there is no reporting URL and nothing is reported.
	h. Start a connection to server 4 which has reporting URL 1, for the 1st time.
		i. I see that a cookie is received from the reporting server 1.
		ii. The received cookie is saved for the server 4.
	i. Start a 2nd connection to server 4.
		i. I see the cookie of server 4 is sent from Outline to the reporting server.
		ii. No cookie is received from the reporting server.
	j. Start a connection to server 1 with reporting URL 1 (the same as server 4).
		i. I see the saved cookie for server 1 is sent from Outline to the reporting server and no cookie is received from the reporting server.
	k. Clear the storage of the Outline app from the device and then run the Outline app.
		i. When adding the same keys and making a connection, I see that no cookie is sent from it to the server which means the cookie file has been deleted. I also verified this by printing the content of the cookie.json file and found out it is empty.

fortuna commented on client/go/outline/client.go:
```diff
@@ -45,7 +48,8 @@ func (c *Client) ListenPacket(ctx context.Context) (net.PacketConn, error) {
 type ClientConfig struct {
-	Transport configyaml.ConfigNode
+	Transport       configyaml.ConfigNode
+	ReportingConfig configyaml.ConfigNode `yaml:"reporting_config,omitempty"`
```
No need for `_config`. Drop the suffix

emohandesi replied:
Currently, we use the tag `reporting_config` in the YAML file to show the reporting section as shown below:
```yaml
reporting_config:
  $type: sessionreport
  url: https://your-callback-server.com/outline_callback
  interval: 24h
  enable_cookies: true
```
Do you mean to change this tag to `reporting` as shown below:
```diff
-reporting_config:
+reporting:
  $type: sessionreport
  url: https://your-callback-server.com/outline_callback
  interval: 24h
  enable_cookies: true
```

fortuna replied:
I think `report` is fine. Also, for fields with `_`, I'd rather have that in the name of the field than use the `yaml` tag, because it's less tied to the yaml library, and we don't want to be tied to a specific library.

fortuna replied:
IT seems the code is already changed to `Report` struct field with `report` config field. That's good, no `_`.

emohandesi commented on client/go/outline/reporting/reporting_client.go:
```diff
@@ -0,0 +1,314 @@
+	configDir, err := os.UserConfigDir()
+	if err != nil {
+		slog.Error("failed to get user config dir", "err", err)
+		configDir = "/data/data/org.outline.android.client/files/"
```
This code is gone now.

jyyi1 replied:
Thanks. Did you push the code? My view still shows the old code.

emohandesi replied:
Sorry, there were a few more changes still remaining. I just pushed.

emohandesi commented on client/config.md:
```diff
@@ -121,6 +121,27 @@ Note that Websockets is not yet supported on Windows. In order to have a single
+      interval: 24h
+      enable_cookies: true`)
+
+    tcpudp with session 
```
Moved this code to a new section, so this code does not exist anymore.

emohandesi committed:
Merge remote-tracking branch 'upstream/master' into usage-reporting-v2

fortuna commented on client/src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -300,13 +303,26 @@ private synchronized PlatformError startTunnel(
+        File filesDir = this.getFilesDir();
+        if (filesDir != null) {
+            this.outlineClient.setCookieFilePath(filesDir.getAbsolutePath());
+            this.outlineClient.startReporting();
+        } else {
+            LOG.log(Level.SEVERE, "Cookies path is null");
+        }
```
Let's move this so it's always called on session start.

```suggestion
        } else {
            LOG.log(Level.SEVERE, "Cookies path is null");
        }
        this.outlineClient.startSession();
```
